pkg/controller/garbagecollector/patch.go

```diff
@@ -52,3 +147,24 @@ func (n *node) patchToUnblockOwnerReferences() ([]byte, error) {
 	dummy.ObjectMeta.UID = n.identity.UID
 	return json.Marshal(dummy)
 }
+
+// Generate a JSON merge patch that unsets the BlockOwnerDeletion field of all
+// ownerReferences of node.
+// NOTE: The unnamed list of UID parameter is needed to implement jsonMergePatchFunc type. The input is not used.
+// This function will operate on all owner references.
+func (gc *GarbageCollector) unblockOwnerReferencesJSONMergePatch(n *node, _ ...types.UID) ([]byte, error) {
+	accessor, err := gc.getCachedMetadata(n.identity.APIVersion, n.identity.Kind, n.identity.Namespace, n.identity.Name)
+	if err != nil {
+		return nil, err
+	}
+	expectedObjectMeta := ObjectMetaForPatch{}
```

Comment:
@deads2k @liggitt do you have better ideas to construct json merge patch? We used `ObjectMetaForPatch` to hold patch contents instead of using `ObjectMeta` directly because some optional fields in `ObjectMeta` is not pointer, so they are serialized as `<field name>: null` instead of being omitted, so the resulting json merge patch would be wrong.

Reply:
> so they are serialized as <field name>: null instead of being omitted, so the resulting json merge patch would be wrong.

seems like you need to ensure the fields are never null, e.g. `expectedOwners := []metav1.OwnerReference{}`

Reply:
which fields were problematic? non-pointer slice fields can be set to null

Reply:
Actually only one problematic field: `CreationTimestamp`.

```golang
	meta := v1.ObjectMeta{}
	meta.OwnerReferences = []v1.OwnerReference{
		{UID: "freaf"},
	}
	meta.ResourceVersion = "1234"
	data, err := json.Marshal(meta)
	if err != nil {
		panic(err)
	}
	fmt.Printf("data=%s\n", data)
```
Output is 
```
data={"resourceVersion":"1234","creationTimestamp":null,"ownerReferences":[{"apiVersion":"","kind":"","name":"","uid":"freaf"}]}
```
Maybe we can manually remove the `creationTimestamp` part, and add a unit test to make sure the patch only contains `resourceVersion` and `ownerReferences`.

Reply:
I'd probably use the unstructured object like this:
```go
u := &unstructured.Unstructured{Object: map[string]interface{}{}}
u.SetResourceVersion(...)
u.SetOwnerReferences(...)
data, err := json.Marshal(u)
```

Reply:
Unstructured object uses reflection which may be slow
https://github.com/kubernetes/kubernetes/blob/ea92879fab9eb9cd7a2529101fb088c2f07bd547/staging/src/k8s.io/apimachinery/pkg/runtime/converter.go#L409-L416

Marshal owner references and construct final patch using `fmt.Sprintf` like what we do for [SMP](https://github.com/kubernetes/kubernetes/blob/b5d21a92adbe829784aa1fb41fa12a4414958f07/pkg/controller/garbagecollector/patch.go#L34) instead

Reply:
The code mentioned above doesn’t use that reflective conversion path. Please don’t use Sprintf to form structured patches that contain any user controlled data.

Reply:
SetOwnerReferences uses the conversion path https://github.com/kubernetes/kubernetes/blob/b5d21a92adbe829784aa1fb41fa12a4414958f07/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L181-L184

Is the concern about using Sprintf that we may change the json field name (`ownerReferences`) in future?

Reply:
The concern is that data won't be escaped or formatted properly. Sprintf isn't a proper tool for marshaling/escaping structured data

Reply:
Did a benchmark for marshaling the patch but it's hard to tell how it will affect GC performance:
```
func BenchmarkReflective(b *testing.B) {
	owners := []v1.OwnerReference{
		{UID: "freaf"},
	}
	resourceVersion := "1234"
	for i := 0; i < b.N; i++ {
		u := &unstructured.Unstructured{Object: map[string]interface{}{}}
		u.SetResourceVersion(resourceVersion)
		u.SetOwnerReferences(owners)
		_, err := json.Marshal(u)
		if err != nil {
			panic(err)
		}
	}
}

func BenchmarkSprintf(b *testing.B) {
	owners := []v1.OwnerReference{
		{UID: "freaf"},
	}
	resourceVersion := "1234"
	for i := 0; i < b.N; i++ {
		expectedOwnersJSON, err := json.Marshal(owners)
		if err != nil {
			panic(err)
		}
		_ = []byte(fmt.Sprintf(`{"metadata":{"resourceVersion":"%s","ownerReferences":%s}}`, resourceVersion, expectedOwnersJSON))
	}
}

Result: 
BenchmarkReflective-12            200000              7368 ns/op
BenchmarkSprintf-12              2000000               894 ns/op
```

Withdrew the last commit to go with the single-use typed struct

Reply:
@liggitt @lavalamp I still prefer using the single-user typed `ObjectMetaForPatch` struct. If the current implementation looks good I can squash.

Some alternatives: 

1. Use json patch instead of jsonmerge patch. The benefit is that the patch size will be smaller when we delete some owner references. The json patch only contains the owner references we want to delete, instead of the entire owner references array in json merge patch. For unblocking owner references json patch still has to include the entire array. If we want to avoid allocation, we can use Sprintf to construct the json patch. The downside is still that Sprintf isn't a proper tool for marshaling/escaping structured data. And json patch may require more careful implementation & testing regarding deleting element in array by indexes.

2. Allocate a map[string]interface{} and marshal it e.g.:
```
metadata := map[string]interface{} {
    "metadata": map[string]interface{} {
        "resourceVersion": string{resourceVersion},
        "ownerReferences": []metav1.OwnerReference {owners},
    },
}
patch, err := json.Marshal(metadata)
```